src/main/scala/translating/BAPToIR.scala

```diff
@@ -77,17 +82,138 @@ class BAPToIR(var program: BAPProgram, mainAddress: BigInt) {
     Program(procedures, mainProcedure.get, memorySections)
   }
 
-  private def translate(s: BAPStatement) = s match {
+  private def translateStatement(s: BAPStatement): Seq[Statement] = s match {
     case b: BAPMemAssign =>
-      val mem = b.lhs.toIRMemory
-      if (mem != b.rhs.memory.toIRMemory) {
+      val mem = translateMemory(b.lhs)
+      if (mem != translateMemory(b.rhs.memory)) {
         throw Exception(s"$b has conflicting lhs ${b.lhs} and rhs ${b.rhs.memory}")
       }
-      MemoryAssign(mem, b.rhs.index.toIR, b.rhs.value.toIR, b.rhs.endian, b.rhs.size, Some(b.line))
+      Seq(MemoryStore(mem, translateExprOnly(b.rhs.index), translateExprOnly(b.rhs.value), b.rhs.endian, b.rhs.size, Some(b.line)))
     case b: BAPLocalAssign =>
-      Assign(b.lhs.toIR, b.rhs.toIR, Some(b.line))
+      val lhs = translateVar(b.lhs)
+      val (rhs, load) = translateExpr(b.rhs)
+      if (load.isDefined) {
+        val loadWithLabel = MemoryLoad(load.get.lhs, load.get.mem, load.get.index, load.get.endian, load.get.size, Some(b.line + "$0"))
+        val assign = LocalAssign(lhs, rhs, Some(b.line + "$1"))
+        Seq(loadWithLabel, assign)
+      } else {
+        val assign = LocalAssign(lhs, rhs, Some(b.line))
+        Seq(assign)
+      }
   }
 
+  private def translateExpr(e: BAPExpr): (Expr, Option[MemoryLoad]) = e match {
+    case BAPConcat(left, right) => (BinaryExpr(BVCONCAT, translateExprOnly(left), translateExprOnly(right)), None)
```

Comment:
It's a bit crude at the moment, yes, mostly to avoid having to consider the case where both sides of a BinaryExpr could contain a load (which the lifter should not produce), since I was unaware of any cases where a Concat could contain a load to begin with. This is easy enough to fix though.